apps/accounts/social_auth_pipeline.py

```diff
@@ -28,99 +40,148 @@
     """
+
     def wrapped(strategy, *args, **kwargs):
         if not kwargs.get('request'):
             request = strategy.request
             kwargs['request'] = request
         return pipeline_func(*args, **kwargs)
+
     return wrapped
 
+
 def reset_session_keys(strategy, *args, **kwargs):
     """Reset a bunch of keys used as part of the social auth flow
-    This is to prevent partially-completed values from a previous flow from affecting a new social auth flow
+
+    This is to prevent partially-completed values from a previous flow from affecting a
+    new social auth flow
     """
     for key in SOCIAL_AUTH_FLOW_KEYS:
         if strategy.request.session.get(key):
             del strategy.request.session[key]
     return None
 
+
 @partial
-def check_email(strategy, details, user=None, *args, **kwargs):
+def check_email(strategy, details, backend, uid, user=None, *args, **kwargs):
     """Ask the user to enter the email if we don't have one yet
 
-    The pipeline process was cut prior to this custom pipeline function, and will resume to this same function after completing
+    The pipeline process was cut prior to this custom pipeline function, and will
+    resume to this same function after completing the social auth flow.
     """
     response = None
     if user is None:
-        strategy.request.session['backend'] = kwargs.get('current_partial').backend
+        strategy.request.session['backend'] = kwargs.get(
+            'current_partial'
+        ).backend
         social_email = details.get('email')
-        collected_email = strategy.request.session.get(SOCIAL_REGISTRATION_SETTING_EMAIL)
+        collected_email = strategy.request.session.get(
+            SOCIAL_REGISTRATION_SETTING_EMAIL
+        )
         if social_email:
             # email available from social auth
             user = get_user_by_email(social_email)
             if user and user.is_active:
-                # a user is already associated with this email
-                # TODO: there is an error with linking accounts...
-                strategy.request.session[SOCIAL_REGISTRATION_SETTING_EMAIL] = social_email
-                if user.has_usable_password():
-                    # user should log into the existing account with a password
-                    url_name = htk_setting('HTK_ACCOUNTS_REGISTER_SOCIAL_LOGIN_URL_NAME')
+                # A user is already associated with this email
+                auto_associate_backends = htk_setting(
+                    'HTK_ACCOUNTS_SOCIAL_AUTO_ASSOCIATE_BACKENDS', []
+                )
+                backend_name = backend.name
+                if backend_name in auto_associate_backends:
+                    # The backend is one of the auto-associate backends, so we need to
+                    # associate the Django user with the social auth account
+                    # This is a job for `associate_user` pipeline but the association
+                    # must happen in here.
```

Comment:
This makes sense. Could we expand this command and include the reasoning `why` ?

"... must happen in here because ..."